### PR TITLE
feat: always send messages in sanic background process

### LINKED ISSUE #189: Send events to redis from a cronjob
Currently we try to send events to redis and also write to the DB in case redis is not available. Then we periodically check if any events were not able to be sent and have accumulated in postgres and try to resend them.

We also have to keep deleting events from postgres that were successfully sent.

We should clean this up and write all events to postgres and then periodically flush them to redis via a cronjob. We should extend the functionality of the existing cronjob that syncs with Keycloak to do this as well.

### PR DESCRIPTION
**Pull request by Panaetius**
closes #189 

/deploy #notest renku=release-0.52.x

### CODE REVIEW

File: bases/renku_data_services/data_api/main.py [hunk 1]
```diff
@@ -26,6 +27,30 @@
...
+            logger.warning(f"Exiting: {e}")
+            return
+        except Exception as e:
+            logger.warning(f"Background task failed: {e}")
+            raise
+
+
+def send_pending_events() -> None:
+    """Send pending messages in case sending in a handler failed."""
+    app = Sanic("send_events")  # we need a dummy app for logging to work.  # noqa: F841
```
Comment (leafty) [inline bases/renku_data_services/data_api/main.py:46]: Would this work?

```suggestion
   Sanic("send_events")  # we need a dummy app for logging to work.
```

Comment (Panaetius) [reply bases/renku_data_services/data_api/main.py:46]: i can give it a try, just copied that from your earlier PR :slightly_smiling_face:

Comment (leafty) [reply bases/renku_data_services/data_api/main.py:46]: 😄 I just pushed the proof of concept for creating a managed background process and how to establish a communication channel. It wasn't production-ready code yet.

My intuition on this is that the logger may want access to `current_app` and that's why the line is needed here, btw.

File: test/bases/renku_data_services/data_api/test_projects.py [hunk 2]
```diff
@@ -69,16 +65,7 @@ async def test_project_creation(sanic_client, user_headers, regular_user, app_co
         "http://renkulab.io/repository-1",
         "http://renkulab.io/repository-2",
     }
-    events = await app_config.redis.redis_connection.xrange("project.created")
-    assert len(events) == 1
-    event = events[0][1]
-    headers = avro_schema_v1.Header.deserialize(event.get(b"headers"), serialization_type="avro-json")
-    assert headers.source == "renku-data-services"
-    proj_event = deserialize_binary(event[b"payload"], avro_schema_v2.ProjectCreated)
-    assert proj_event.name == payload["name"]
```
[the remark this answers is not in the document]
Comment (olevski) [reply test/bases/renku_data_services/data_api/test_projects.py:78]: This also applies to all the other tests where these kind of checks are removed. Considering we are worried about drifting from the events queue and all its consumers we should test that we are adding the right events in all cases.

**Review by olevski — CHANGES_REQUESTED** [review]
Thanks Ralf just a few minor changes.

File: bases/renku_data_services/data_api/main.py [hunk 3]
```diff
@@ -26,6 +27,30 @@
...
+
+    logger.info("running events sending loop.")
+
+    asyncio.set_event_loop(uvloop.new_event_loop())
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply bases/renku_data_services/data_api/main.py:50]: It's what sanic uses internally as well and I didn't want to deviate from that

File: bases/renku_data_services/data_api/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Panaetius) [reply bases/renku_data_services/data_api/main.py:46]: Right below that line it says: `The simplest method for doing that in Python will be to just wrap your loop in KeyboardInterrupt` which I added here: https://github.com/SwissDataScienceCenter/renku-data-services/pull/217/files#diff-1b9249f808cb3fb3ed6b2320c83439f3efcaca58faf9f93bb52e6ad03e1b301dR36

File: components/renku_data_services/message_queue/db.py [hunk 4]
```diff
@@ -26,36 +25,43 @@ def __init__(
...
-            return
-        for event in events_orm:
-            try:
-                await self.message_queue.send_message(event.dump())
-                await self.delete_event(event.id)
-            except Exception as e:
-                logger.warning(f"couldn't resend event {event.payload} on queue {event.queue}: {e}")
-
-        logger.info(f"resent {num_events} events")
+        while True:
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply components/renku_data_services/message_queue/db.py:40]: Maybe I misunderstand, but there is an asyncio.sleep 3 lines below

File: test/components/renku_data_services/message_queue/test_queue.py [hunk 5]
```diff
@@ -48,33 +34,21 @@ async def fake_db_method(self, some_arg, *, session: AsyncSession | None = None)
...
-
-    events = await app_config.redis.redis_connection.xrange("project.removed")
-    assert len(events) == 0
+    assert len(events) == 1
     pending_events = await app_config.event_repo._get_pending_events()
-    assert len(pending_events) == 1
+    assert len(pending_events) == 0
 
     # ensure it is resent if older than 5 seconds
-    await asyncio.sleep(6)
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply test/components/renku_data_services/message_queue/test_queue.py:77]: yes I missed that :+1:

File: bases/renku_data_services/data_api/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (olevski) [reply bases/renku_data_services/data_api/main.py:46]: Sorry I missed that. Its all good then.

File: components/renku_data_services/message_queue/db.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (olevski) [reply components/renku_data_services/message_queue/db.py:40]: This is what I see in the PR review and there is no sleep:
image

**Pull request merged**